---
title: Java - Exception
date: 2023-10-14 13:00:00 +0900
categories: [Java]
published: true
tags:
- Java
- Exception
---

### 예외 계층
![Alt text](/assets/posts/img/java/java_01_01.png)
  - Object
    - 예외도 객체
    - 모든 객체의 최상위 부모는 Object이므로 예외의 최상위 부모도 Object
  - Throwable
    - 최상위의 예외
  - Error
    - 메모리 부족이나 심각한 시스템 오류와 같이 어플리케이션에서 복구 불가능한 시스템 예외
    - 어플리케이션 개발자는 이 예외를 잡으면 안됨
      - 상위 예외를 catch로 잡으면 그 하위 예외까지 잡히기 떄문에 Throwable 예외도 잡으면 안됨
        - Throwable 예외를 catch로 잡으면 하위인 Error 예외도 함꼐 잡히기 때문
    - 언체크 예외
  - Exception
    - 체크 예외
    - 어플리케이션 로직에서 사용할 수 있는 실질적인 최상위 예외
    - Exception와 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외임. 단 RuntimeException만은 해당이 안됨
  - RuntimeException
    - 언체크 예외. 런타임 예외
    - 컴파칠러가 체크하지 않는 언체크 예외
    - RuntimeException과 그 자식 예외는 모두 언체크 예외 임
    - RuntimeException의 이름을 따라서 RuntimeException과 그 하위 언체크 예외를 런타임 에러라고도 함

### 예외의 기본 규칙
예외는 기본적으로 잡아서 처리하거나, 처리할 수 없다면 밖으로 던져야 함
  - 예외는 잡아서 처리하거나 던져야 함
  - 예외를 잡거나 던질떄 지정한 예외 뿐만 아니라 그 예외의 자식들도 함꼐 처리됨
    - 예를 들어 Exception을 catch로 잡으면 그 하위 예외들도 모두 잡을 수 있음
    - 예를 들어 Exception을 throws로 던지면 그 하위 예외들도 모두 던질 수 있음

**예외 처리**
![Alt text](/assets/posts/img/java/java_01_02.png)
  - 5번에서 예외를 처리하면 이후에는 어플리케이션 로직이 정상흐름으로 동작

**예외 던짐**
![Alt text](/assets/posts/img/java/java_01_03.png)
  - 여예를 처리하지 못하면 호출한 곳으로 예외를 계속 던짐 

#### 체크 예외
 - Exception과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외. 단 RuntimeException만은 제외됨
 - 체크 예외는 잡아서 처리하거나 또는 밖으로 던지도록 선언해야 함. 그렇지 않으면 컴파일 오휴 발생

```java
static class MyCheckedException extends Exception {
    public MyCheckedException(String message) {
        super(message);
    }
}
```
```java
// 체크 예외를 잡아서 처리하는 코드
try {
    repository.call();
} catch (MyCheckedException e) {
    //예외 처리 로직
}
```
```java
// 체크 예외를 잡아서 처리하는 코드
try {
    repository.call();
} catch (Exception e) {
    //예외 처리 로직
}
```
  - catch에 예외를 지정하면 해당 예외와 그 하위 타입 예외를 모두 잡음
    - MyCheckedException의 상위 타입인 Exception을 적더라도 MyCheckedException을 잡을 수 있음

```java
// 체크 예외를 밖으로 던지는 코드
public void callThrow() throws MyCheckedException {
    repository.call();
}
```
```java
public void callThrow() throws Exception {
    repository.call();
}
```
  - 체크 예외를 처리할 수 없는 경우에는 "Method() throws 예외"를 사용하여 밖으로 던질 예외를 필수적으로 지정해야 함
  - throw를 지정하지 않으면 첨파일 오류 발생
  - 체크 예외의 경우 예외를 잡아서 처리하거나 또는 throws를 지정하여 예외를 밖으로 던진다는 선언을 필수로 해주어야 함
  - throws에 지전한 타입과 그 하위 타입 예외를 밖으로 던짐
    - throws에 MyCheckedException의 상위 타입인 Exception을 적어주어도 MyCheckedException을 던짐

#### 언처크 예외
 - RuntimeException과 그 하위 예외는 언체크 예외로 분류
 - 언체크 예외는 컴파일러가 예외를 체크하지 않는다는 뜻
 - 언체크 예외는 체크 예외와 기본적으로 동일하지만 언체크 예외는 예외를 던지는 throws를 선언하지 않고 생략이 가능함. 이 경우 자동으로 예외를 던짐

#### 체크 예외 vs 언체크 예외
  - 체크 예외 : 예외를 잡아서 처리하지 않으면 항상 throws에 던지는 예외 선언을 해야함
    - 해당 예외를 잡아서 반드시 처리해야하는 비지니스 로직상의 문제에서 던지는 예외에만 사용
      - 예를 들면 계좌 이체 실패, 결제시 포인틑 부족, 로그인 ID, PW 불필치 등
  - 언체크 예외 : 예외를 잡아서 처리하지 않아도 throws를 생략할 수 있음
    - 기본적으로 언체크 예외를 사용하는 것이 좋음 

사실상 체크 예외와 언체크 예외의 차이는 예외를 처리할 수 없을 때 예외를 밖으로 던지는 부분에 있음.
throws 부분을 필수로 선언해야 하는가 생략할 수 있는가의 차이임.

### 체크 예외의 문제점
![Alt text](/assets/posts/img/java/java_01_04.png)
  - 서비스
    - 체크 예외를 처리하지 못해서 밖으로 던지기 위해  logic() throws SQLException, ConnectException 를 선언
  - 컨트롤러
    - 체크 예외를 처리하지 못해서 밖으로 던지기 위해  request() throws SQLException, ConnectException 를 선언
 
위 그림으로 볼떄 크게 두가지 문제가 있을 수 있음
  - 복구 불가능한 예외
  - 의존 관계에 대한 문제

#### 복구 불가능한 예외
대부준의 예외는 복구가 불가능 함. SQLException을 예로 들면 데이터베이스에 무언가 문제가 있어서 발생하는 예외인데, SQL 문법에 문제가 있을 수 있고, 데이터베이스 자체에 문제가 발생했을 수도 있으며 데이터베이스가 아예 다운되었을 수도 있음. 이러한 문제는 대체로 어플리케이션에선 복구가 불가능 함. 따라서 대부분의 서비스와 컨트롤러에서도 이 문제를 해결 할 수 없으며 해결하지 못하는 예외를 공통 예외 처리 로직 까지 예외를 던져야 함 

#### 의존 관계에 대한 문제
체크 예외이기 때문에 서비스나 컨트롤러 입장에서는 자신이 처리하지 못하더라도 throws를 통해 예외를 던져야 함

```java
class Controller {
    public void request() throws SQLException, ConnectException {
        service.logic();
    }
}
class Service {
    public void logic() throws SQLException, ConnectException {
        repository.call();
        networkClient.call();
    }
}
```

위와 같은 결루, 서비스나 컨트롤러는 단순히 예외를 던지기 위해 SQLException를 의존하게 되는 문제가 발생 함.
추후 리포지토리를 JDBC 기술이 아닌 다른 기술로 변경해야 한다면 SQLException에 의존하던 모든 서비스, 컨트롤러의 코드를 수정해야 함.
결국 자신이 처리하지도 못하는 예외로 인하여 의존관계가 성립되어 OCP, DI에 위배되는 코드가 되는 문제가 있음

### 런타임 예외의 활용
![Alt text](/assets/posts/img/java/java_01_05.png)
  - SQLException을 런타임 예외인 RuntimeSQLException으로 변환
  - ConnectionExceoption 대신 RuntimeConnectionExceoption으로 변환
  - 런타임 예이이기 때문에 서비스, 컨트롤러는 해당 예외를 처리할 수 없다면 별도의 선언 없이 그냥 두면 됨

#### 런타임 예외 - 복구 불가능한 예외
시스템에서 발생하는 예외는 대부분 복구 불가능한 예외.
런타임 예외를 사용하면 서비스나 컨트롤러가 이러한 복구 불가능한 예외를 신경쓰지 않아도 됨. 물론 이러한 복구 불가능한 예외는 일관성 있게 공통으로 처리해야 함

#### 런타임 예외 - 의존 관계에 대한 문제
런타임 예외는 해당 객체가 처리할 수 없는 예외는 무시하면 됨. 그러므로 체크 예외 처럼 throws를 갈제로 사용하지 않아도 되기 때문에 예외를 강제로 의존하지 않음

![Alt text](/assets/posts/img/java/java_01_06.png)
  - 서비스와 컨트롤러에선 의존관계가 발생하지 않기 때문에 기술 변경시 코드를 변경하지 않아도 됨
  - 공통으로 예외를 처리하는 부분의 코드만 변경하면 되기 때문에 변경의 영향 범위를 최소화 할 수 있음

### 예외 포함
예외를 전환할떄는 발생한 예외의 원일 파악을 위해 반드시 기존 예외가 포홈되어야 함

```java
public void call() {
    try {
        runSQL();
    } catch (SQLException e) {
        throw new RuntimeSQLException(e); //기존 예외(e) 포함
    }
}
```
```
13:10:45.626 [Test worker] INFO hello.jdbc.exception.basic.UncheckedAppTest - 
ex
hello.jdbc.exception.basic.UncheckedAppTest$RuntimeSQLException: 
java.sql.SQLException: ex
at 
hello.jdbc.exception.basic.UncheckedAppTest$Repository.call(UncheckedAppTest.ja
va:61)
at 
hello.jdbc.exception.basic.UncheckedAppTest$Service.logic(UncheckedAppTest.java
:45)
at 
hello.jdbc.exception.basic.UncheckedAppTest$Controller.request(UncheckedAppTest
.java:35)
at 
hello.jdbc.exception.basic.UncheckedAppTest.printEx(UncheckedAppTest.java:24)
Caused by: java.sql.SQLException: ex
at 
hello.jdbc.exception.basic.UncheckedAppTest$Repository.runSQL(UncheckedAppTest.
java:66)
at 
hello.jdbc.exception.basic.UncheckedAppTest$Repository.call(UncheckedAppTest.ja
va:59)
```
```java
public void call() {
    try {
        runSQL();
    } catch (SQLException e) {
        throw new RuntimeSQLException(); //기존 예외(e) 미포함
    }
}
```
```
[Test worker] INFO hello.jdbc.exception.basic.UncheckedAppTest - ex
hello.jdbc.exception.basic.UncheckedAppTest$RuntimeSQLException: null
at 
hello.jdbc.exception.basic.UncheckedAppTest$Repository.call(UncheckedAppTest.ja
va:61)
at 
hello.jdbc.exception.basic.UncheckedAppTest$Service.logic(UncheckedAppTest.java
:45)
```
  - 기존 예외를 포함하지 않아서 기존에 발생한 SQLException과 스텍 트레이스를 확인 할 수 없고 변환한 RuntimeSQLException부터 예외를 확인할 수 있음

**예외를 전환할 때는 반드시 기존 예외를 포함하도록 해야 함**

### 참고
 - 스프링 DB - 데이터 접근 핵심 원리(김영한)
